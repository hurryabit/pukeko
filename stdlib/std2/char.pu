-- ASCII characters
import std2/int

type Char = external

ord :: Char -> Int
external ord = "ord"

-- Does not check for overflow, just takes the least significant byte.
chr :: Int -> Char
external chr = "chr"
