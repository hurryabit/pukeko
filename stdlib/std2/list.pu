-- The List datatype
import std2/basic
import std2/bool
import std2/foldable
import std2/functor
import std2/monoid

data List a = Nil | Cons a (List a)

instance eqList :: (Eq a) => Eq (List a) where
  eq xs ys = case xs of
    Nil -> case ys of
      Nil -> True
      Cons _ _ -> False
    Cons x xs -> case ys of
      Nil -> False
      Cons y ys -> x == y && xs == ys

instance monoidList :: Monoid (List a) where
  empty = Nil
  append xs ys = foldr Cons ys xs

instance functorList :: Functor List where
  map f xs = case xs of
    Nil -> Nil
    Cons x xs -> Cons (f x) (map f xs)

instance foldableList :: Foldable List where
  foldr f y0 xs = case xs of
    Nil -> y0
    Cons x xs -> f x (foldr f y0 xs)
  foldl f y0 xs = case xs of
    Nil -> y0
    Cons x xs -> foldl f (f y0 x) xs

to_list :: (Foldable t) => t a -> List a
to_list = foldr Cons Nil

nth_exn :: List a -> Int -> a
nth_exn xs n =
  case xs of
    Nil -> abort
    Cons x xs -> if n <= 0 then x else nth_exn xs (n-1)

take :: Int -> List a -> List a
take n xs =
  if n <= 0 then
    Nil
  else
    case xs of
      Nil       -> Nil
      Cons x xs -> Cons x (take (n-1) xs)

replicate :: Int -> a -> List a
replicate n x =
  if n <= 0 then Nil else Cons x (replicate (n-1) x)

zip_with :: (a -> b -> c) -> List a -> List b -> List c
zip_with f xs ys =
  case xs of
    Nil -> Nil
    Cons x xs ->
      case ys of
        Nil -> Nil
        Cons y ys -> Cons (f x y) (zip_with f xs ys)

partition :: (a -> Bool) -> List a -> Pair (List a) (List a)
partition p xs =
  case xs of
    Nil -> Pair Nil Nil
    Cons x xs ->
      case partition p xs of
        Pair ys zs ->
          if p x then
            Pair (Cons x ys) zs
          else
            Pair ys (Cons x zs)
