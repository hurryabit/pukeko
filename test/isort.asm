g_declare_cafs Nil, prime, powers, sum, numbers, main
g_declare_main main

g_globstart prefix_add, 2
g_push 1
g_eval
g_push 1
g_eval
g_add
g_update 3
g_pop 2
g_return

g_globstart prefix_sub, 2
g_push 1
g_eval
g_push 1
g_eval
g_sub
g_update 3
g_pop 2
g_return

g_globstart prefix_mul, 2
g_push 1
g_eval
g_push 1
g_eval
g_mul
g_update 3
g_pop 2
g_return

g_globstart Nil, 0
g_cons 0, 0
g_update 1
g_return

g_globstart print, 2
g_eval
g_print
g_cons 0, 0
g_cons 0, 2
g_update 1
g_return

g_globstart foldr, 3
g_push 2
g_eval
g_jumpzero .0
g_uncons 2
g_push 1
g_push 4
g_push 4
g_pushglobal foldr, 3
g_mkap 3
g_push 1
g_push 4
g_mkap 2
g_update 6
g_pop 5
g_unwind
g_jump .1
g_label .0
g_uncons 0
g_push 1
g_eval
g_update 4
g_pop 3
g_unwind
g_label .1

g_globstart take, 2
g_pushint 0
g_push 1
g_eval
g_leq
g_jumpzero .0
g_uncons 0
g_pushglobal Nil, 0
g_eval
g_update 3
g_pop 2
g_unwind
g_jump .1
g_label .0
g_uncons 0
g_push 1
g_eval
g_jumpzero .2
g_uncons 2
g_push 1
g_pushint 1
g_push 4
g_pushglobal prefix_sub, 2
g_mkap 2
g_pushglobal take, 2
g_mkap 2
g_push 1
g_cons 1, 2
g_update 5
g_pop 4
g_return
g_jump .3
g_label .2
g_uncons 0
g_pushglobal Nil, 0
g_eval
g_update 3
g_pop 2
g_unwind
g_label .3
g_label .1

g_globstart zip_with, 3
g_push 1
g_eval
g_jumpzero .0
g_uncons 2
g_push 4
g_eval
g_jumpzero .2
g_uncons 2
g_push 1
g_push 4
g_push 6
g_pushglobal zip_with, 3
g_mkap 3
g_push 1
g_push 4
g_push 7
g_mkap 2
g_cons 1, 2
g_update 8
g_pop 7
g_return
g_jump .3
g_label .2
g_uncons 0
g_pushglobal Nil, 0
g_eval
g_update 6
g_pop 5
g_unwind
g_label .3
g_jump .1
g_label .0
g_uncons 0
g_pushglobal Nil, 0
g_eval
g_update 4
g_pop 3
g_unwind
g_label .1

g_globstart prime, 0
g_pushint 39
g_pushint 1000000
g_pushint 1000000
g_mul
g_add
g_update 1
g_pop 0
g_return

g_globstart double_mod_prime, 1
g_push 0
g_pushint 2
g_pushglobal prefix_mul, 2
g_mkap 2
g_pushglobal prime, 0
g_eval
g_push 1
g_eval
g_les
g_jumpzero .0
g_uncons 0
g_push 0
g_eval
g_update 3
g_pop 2
g_unwind
g_jump .1
g_label .0
g_uncons 0
g_pushglobal prime, 0
g_eval
g_push 1
g_eval
g_sub
g_update 3
g_pop 2
g_return
g_label .1

g_globstart gen, 2
g_push 1
g_push 1
g_mkap 1
g_push 1
g_pushglobal gen, 2
g_mkap 2
g_push 2
g_cons 1, 2
g_update 3
g_pop 2
g_return

g_globstart powers, 0
g_pushint 1
g_pushglobal double_mod_prime, 1
g_pushglobal gen, 2
g_mkap 2
g_update 1
g_pop 0
g_unwind

g_globstart sum, 0
g_pushint 0
g_pushglobal prefix_add, 2
g_pushglobal foldr, 3
g_mkap 2
g_update 1
g_pop 0
g_unwind

g_globstart mul_mod_prime, 2
g_pushglobal prime, 0
g_eval
g_push 2
g_eval
g_push 2
g_eval
g_mul
g_mod
g_update 3
g_pop 2
g_return

g_globstart sum_prod, 2
g_push 1
g_push 1
g_pushglobal mul_mod_prime, 2
g_pushglobal zip_with, 3
g_mkap 3
g_pushglobal sum, 0
g_mkap 1
g_update 3
g_pop 2
g_unwind

g_globstart hash, 1
g_pushglobal prime, 0
g_eval
g_pushglobal powers, 0
g_push 2
g_pushglobal sum_prod, 2
g_mkap 2
g_eval
g_mod
g_update 2
g_pop 1
g_return

g_globstart numbers$1, 1
g_pushint 10007
g_push 1
g_eval
g_pushint 42
g_mul
g_mod
g_update 2
g_pop 1
g_return

g_globstart numbers, 0
g_pushint 1
g_pushglobal numbers$1, 1
g_pushglobal gen, 2
g_mkap 2
g_pushint 10006
g_pushglobal take, 2
g_mkap 2
g_update 1
g_pop 0
g_unwind

g_globstart insert, 2
g_push 1
g_eval
g_jumpzero .0
g_uncons 2
g_push 0
g_eval
g_push 3
g_eval
g_leq
g_jumpzero .2
g_uncons 0
g_push 3
g_push 3
g_cons 1, 2
g_update 5
g_pop 4
g_return
g_jump .3
g_label .2
g_uncons 0
g_push 1
g_push 3
g_pushglobal insert, 2
g_mkap 2
g_push 1
g_cons 1, 2
g_update 5
g_pop 4
g_return
g_label .3
g_jump .1
g_label .0
g_uncons 0
g_pushglobal Nil, 0
g_push 1
g_cons 1, 2
g_update 3
g_pop 2
g_return
g_label .1

g_globstart isort, 1
g_push 0
g_eval
g_jumpzero .0
g_uncons 2
g_push 1
g_pushglobal isort, 1
g_mkap 1
g_push 1
g_pushglobal insert, 2
g_mkap 2
g_update 4
g_pop 3
g_unwind
g_jump .1
g_label .0
g_uncons 0
g_pushglobal Nil, 0
g_eval
g_update 2
g_pop 1
g_unwind
g_label .1

g_globstart main, 0
g_pushglobal numbers, 0
g_pushglobal isort, 1
g_mkap 1
g_pushglobal hash, 1
g_mkap 1
g_pushglobal print, 2
g_mkap 1
g_update 1
g_pop 0
g_unwind
