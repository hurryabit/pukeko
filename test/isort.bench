benchmarking ./isort < sort20.in > /dev/null
time                 9.863 s    (9.743 s .. 10.07 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 9.869 s    (9.843 s .. 9.888 s)
std dev              28.14 ms   (0.0 s .. 32.20 ms)
variance introduced by outliers: 19% (moderately inflated)

Steps      :  46640191860
Alloc bytes:  16708800184          (Checks:    596900017)
GC runs    :          125
