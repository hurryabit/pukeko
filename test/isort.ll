external (-) : Int -> Int -> Int = "sub"
external (<=) : Int -> Int -> Bool = "le"
let foldr$ll1 : ∀a b. (a -> b -> b) -> b -> List a -> b =
      fun @a @b ->
        fun (f : a -> b -> b) (y0 : b) (xs : List a) ->
          match xs with
          | Nil @a -> y0
          | Cons @a x xs -> f x (foldr @a @b f y0 xs)
let foldr : ∀a b. (a -> b -> b) -> b -> List a -> b =
      fun @a @b -> foldr$ll1 @a @b
let replicate$ll1 : ∀a. Int -> a -> List a =
      fun @a ->
        fun (n : Int) (x : a) ->
          match (<=) n 0 with
          | False -> Cons @a x (replicate @a ((-) n 1) x)
          | True -> Nil @a
let replicate : ∀a. Int -> a -> List a = fun @a -> replicate$ll1 @a
external return : ∀a. a -> IO a = "return"
external print : Int -> IO Unit = "print"
external input : IO Int = "input"
external (>>=) : ∀a b. IO a -> (a -> IO b) -> IO b = "bind"
let (;ll1) : ∀a. IO a -> Unit -> IO a =
      fun @a -> fun (m2 : IO a) (x : Unit) -> m2
let (;ll2) : ∀a. IO Unit -> IO a -> IO a =
      fun @a ->
        fun (m1 : IO Unit) (m2 : IO a) -> (>>=) @Unit @a m1 ((;ll1) @a m2)
let (;) : ∀a. IO Unit -> IO a -> IO a = fun @a -> (;ll2) @a
let sequence_io$ll1 : ∀a. a -> List a -> IO (List a) =
      fun @a ->
        fun (x : a) (xs : List a) -> return @(List a) (Cons @a x xs)
let sequence_io$ll2 : ∀a. List (IO a) -> a -> IO (List a) =
      fun @a ->
        fun (ms : List (IO a)) (x : a) ->
          (>>=) @(List a) @(List a) (sequence_io @a ms) (sequence_io$ll1 @a x)
let sequence_io$ll3 : ∀a. List (IO a) -> IO (List a) =
      fun @a ->
        fun (ms : List (IO a)) ->
          match ms with
          | Nil @(IO a) -> return @(List a) (Nil @a)
          | Cons @(IO a) m ms -> (>>=) @a @(List a) m (sequence_io$ll2 @a ms)
let sequence_io : ∀a. List (IO a) -> IO (List a) =
      fun @a -> sequence_io$ll3 @a
let iter_io$ll1 : ∀a. (a -> IO Unit) -> a -> IO Unit -> IO Unit =
      fun @a ->
        fun (f : a -> IO Unit) (x : a) (m : IO Unit) -> (;) @Unit (f x) m
let iter_io$ll2 : ∀a. (a -> IO Unit) -> List a -> IO Unit =
      fun @a ->
        fun (f : a -> IO Unit) ->
          foldr @a @(IO Unit) (iter_io$ll1 @a f) (return @Unit Unit)
let iter_io : ∀a. (a -> IO Unit) -> List a -> IO Unit =
      fun @a -> iter_io$ll2 @a
let insert$ll1 : Int -> List Int -> List Int =
      fun (y : Int) (xs : List Int) ->
        match xs with
        | Nil @Int -> Cons @Int y (Nil @Int)
        | Cons @Int x xs' ->
          match (<=) y x with
          | False -> Cons @Int x (insert y xs')
          | True -> Cons @Int y xs
let insert : Int -> List Int -> List Int = insert$ll1
let isort$ll1 : List Int -> List Int =
      fun (xs : List Int) ->
        match xs with
        | Nil @Int -> Nil @Int
        | Cons @Int x xs -> insert x (isort xs)
let isort : List Int -> List Int = isort$ll1
let main$ll1 : List Int -> IO Unit =
      fun (xs : List Int) -> iter_io @Int print (isort xs)
let main$ll2 : Int -> IO Unit =
      fun (n : Int) ->
        (>>=) @(List Int) @Unit (sequence_io @Int (replicate @(IO Int) n input)) main$ll1
let main : IO Unit = (>>=) @Int @Unit input main$ll2
