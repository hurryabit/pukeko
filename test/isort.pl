data Unit =
       | Unit
data Bool =
       | False
       | True
data Pair a b =
       | Pair a b
data Option a =
       | None
       | Some a
data Choice a b =
       | First a
       | Second b
data Dict$Eq a =
       | Dict$Eq (a -> a -> Bool)
data Dict$Ord a =
       | Dict$Ord (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool) (a -> a -> Bool)
le : ∀a. Dict$Ord a -> a -> a -> Bool =
  fun @a ->
    fun (dict : Dict$Ord a) ->
      match dict with
      | Dict$Ord @a _ _ le _ -> le
data Dict$Monoid m =
       | Dict$Monoid m (m -> m -> m)
data Dict$Ring a =
       | Dict$Ring (a -> a) (a -> a -> a) (a -> a -> a) (a -> a -> a)
sub : ∀a. Dict$Ring a -> a -> a -> a =
  fun @a ->
    fun (dict : Dict$Ring a) ->
      match dict with
      | Dict$Ring @a _ _ sub _ -> sub
data Int
external lt_int : Int -> Int -> Bool = "lt"
external le_int : Int -> Int -> Bool = "le"
external ge_int : Int -> Int -> Bool = "ge"
external gt_int : Int -> Int -> Bool = "gt"
dict$Ord$Int : Dict$Ord Int =
  let ge : Int -> Int -> Bool = ge_int
  and gt : Int -> Int -> Bool = gt_int
  and le : Int -> Int -> Bool = le_int
  and lt : Int -> Int -> Bool = lt_int
  in
  Dict$Ord @Int ge gt le lt
external neg_int : Int -> Int = "neg"
external add_int : Int -> Int -> Int = "add"
external sub_int : Int -> Int -> Int = "sub"
external mul_int : Int -> Int -> Int = "mul"
dict$Ring$Int : Dict$Ring Int =
  let neg : Int -> Int = neg_int
  and add : Int -> Int -> Int = add_int
  and sub : Int -> Int -> Int = sub_int
  and mul : Int -> Int -> Int = mul_int
  in
  Dict$Ring @Int neg add sub mul
data Char
data Dict$Foldable t =
       | Dict$Foldable (∀a b. (a -> b -> b) -> b -> t a -> b) (∀a b. (b -> a -> b) -> b -> t a -> b)
foldr : ∀t. Dict$Foldable t -> (∀a b. (a -> b -> b) -> b -> t a -> b) =
  fun @t ->
    fun (dict : Dict$Foldable t) ->
      match dict with
      | Dict$Foldable @t foldr _ -> foldr
foldl : ∀t. Dict$Foldable t -> (∀a b. (b -> a -> b) -> b -> t a -> b) =
  fun @t ->
    fun (dict : Dict$Foldable t) ->
      match dict with
      | Dict$Foldable @t _ foldl -> foldl
data Dict$Functor f =
       | Dict$Functor (∀a b. (a -> b) -> f a -> f b)
data List a =
       | Nil
       | Cons a (List a)
dict$Foldable$List$ll1 : ∀a b. (a -> b -> b) -> b -> List a -> b =
  fun @a @b ->
    fun (f : a -> b -> b) (y0 : b) (xs : List a) ->
      match xs with
      | Nil @a -> y0
      | Cons @a x xs ->
        f x (foldr @List dict$Foldable$List @a @b f y0 xs)
dict$Foldable$List$ll2 : ∀a b. (b -> a -> b) -> b -> List a -> b =
  fun @a @b ->
    fun (f : b -> a -> b) (y0 : b) (xs : List a) ->
      match xs with
      | Nil @a -> y0
      | Cons @a x xs ->
        foldl @List dict$Foldable$List @a @b f (f y0 x) xs
dict$Foldable$List : Dict$Foldable List =
  let foldr : ∀a b. (a -> b -> b) -> b -> List a -> b =
        fun @a @b -> dict$Foldable$List$ll1 @a @b
  and foldl : ∀a b. (b -> a -> b) -> b -> List a -> b =
        fun @a @b -> dict$Foldable$List$ll2 @a @b
  in
  Dict$Foldable @List foldr foldl
replicate : ∀a. Int -> a -> List a =
  fun @a ->
    fun (n : Int) (x : a) ->
      match le @Int dict$Ord$Int n 0 with
      | False -> Cons @a x (replicate @a (sub @Int dict$Ring$Int n 1) x)
      | True -> Nil @a
data Dict$Monad m =
       | Dict$Monad (∀a. a -> m a) (∀a b. m a -> (a -> m b) -> m b)
pure : ∀m. Dict$Monad m -> (∀a. a -> m a) =
  fun @m ->
    fun (dict : Dict$Monad m) ->
      match dict with
      | Dict$Monad @m pure _ -> pure
bind : ∀m. Dict$Monad m -> (∀a b. m a -> (a -> m b) -> m b) =
  fun @m ->
    fun (dict : Dict$Monad m) ->
      match dict with
      | Dict$Monad @m _ bind -> bind
semi$ll1 : ∀a m. m a -> Unit -> m a =
  fun @a @m -> fun (m2 : m a) (x : Unit) -> m2
semi : ∀a m. Dict$Monad m -> m Unit -> m a -> m a =
  fun @a @m ->
    fun (dict$Monad$m : Dict$Monad m) (m1 : m Unit) (m2 : m a) ->
      bind @m dict$Monad$m @Unit @a m1 (semi$ll1 @a @m m2)
sequence$ll1 : ∀a m. Dict$Monad m -> a -> List a -> m (List a) =
  fun @a @m ->
    fun (dict$Monad$m : Dict$Monad m) (x : a) (xs : List a) ->
      pure @m dict$Monad$m @(List a) (Cons @a x xs)
sequence$ll2 : ∀a m. Dict$Monad m -> List (m a) -> a -> m (List a) =
  fun @a @m ->
    fun (dict$Monad$m : Dict$Monad m) (ms : List (m a)) (x : a) ->
      bind @m dict$Monad$m @(List a) @(List a) (sequence @a @m dict$Monad$m ms) (sequence$ll1 @a @m dict$Monad$m x)
sequence : ∀a m. Dict$Monad m -> List (m a) -> m (List a) =
  fun @a @m ->
    fun (dict$Monad$m : Dict$Monad m) (ms : List (m a)) ->
      match ms with
      | Nil @(m a) -> pure @m dict$Monad$m @(List a) (Nil @a)
      | Cons @(m a) m ms ->
        bind @m dict$Monad$m @a @(List a) m (sequence$ll2 @a @m dict$Monad$m ms)
traverse_$ll1 : ∀a m. Dict$Monad m -> (a -> m Unit) -> a -> m Unit -> m Unit =
  fun @a @m ->
    fun (dict$Monad$m : Dict$Monad m) (f : a -> m Unit) (x : a) (m : m Unit) ->
      semi @Unit @m dict$Monad$m (f x) m
traverse_ : ∀a m t. Dict$Monad m -> Dict$Foldable t -> (a -> m Unit) -> t a -> m Unit =
  fun @a @m @t ->
    fun (dict$Monad$m : Dict$Monad m) (dict$Foldable$t : Dict$Foldable t) (f : a -> m Unit) ->
      foldr @t dict$Foldable$t @a @(m Unit) (traverse_$ll1 @a @m dict$Monad$m f) (pure @m dict$Monad$m @Unit Unit)
external seq : ∀a b. a -> b -> b = "seq"
external puti : Int -> Unit = "puti"
external geti : Unit -> Int = "geti"
data World =
       | World
data IO a = World -> Pair a World
dict$Monad$IO$ll1 : ∀a. a -> World -> Pair a World =
  fun @a -> Pair @a @World
dict$Monad$IO$ll2 : ∀a. a -> IO a =
  fun @a ->
    fun (x : a) ->
      coerce @((World -> Pair a World) -> IO a) (dict$Monad$IO$ll1 @a x)
dict$Monad$IO$ll3 : ∀a b. IO a -> (a -> IO b) -> World -> Pair b World =
  fun @a @b ->
    fun (mx : IO a) (f : a -> IO b) (world0 : World) ->
      match coerce @(IO a -> World -> Pair a World) mx world0 with
      | Pair @a @World x world1 ->
        coerce @(IO b -> World -> Pair b World) (f x) world1
dict$Monad$IO$ll4 : ∀a b. IO a -> (a -> IO b) -> IO b =
  fun @a @b ->
    fun (mx : IO a) (f : a -> IO b) ->
      coerce @((World -> Pair b World) -> IO b) (dict$Monad$IO$ll3 @a @b mx f)
dict$Monad$IO : Dict$Monad IO =
  let pure : ∀a. a -> IO a = fun @a -> dict$Monad$IO$ll2 @a
  and bind : ∀a b. IO a -> (a -> IO b) -> IO b =
        fun @a @b -> dict$Monad$IO$ll4 @a @b
  in
  Dict$Monad @IO pure bind
io$ll1 : ∀a b. (a -> b) -> a -> World -> Pair b World =
  fun @a @b ->
    fun (f : a -> b) (x : a) (world : World) ->
      let y : b = f x in
      seq @b @(Pair b World) y (Pair @b @World y world)
io : ∀a b. (a -> b) -> a -> IO b =
  fun @a @b ->
    fun (f : a -> b) (x : a) ->
      coerce @((World -> Pair b World) -> IO b) (io$ll1 @a @b f x)
print : Int -> IO Unit = fun (n : Int) -> io @Int @Unit puti n
input : IO Int = io @Unit @Int geti Unit
insert : Int -> List Int -> List Int =
  fun (y : Int) (xs : List Int) ->
    match xs with
    | Nil @Int -> Cons @Int y (Nil @Int)
    | Cons @Int x xs' ->
      match le @Int dict$Ord$Int y x with
      | False -> Cons @Int x (insert y xs')
      | True -> Cons @Int y xs
isort : List Int -> List Int =
  fun (xs : List Int) ->
    match xs with
    | Nil @Int -> Nil @Int
    | Cons @Int x xs -> insert x (isort xs)
main$ll1 : List Int -> IO Unit =
  fun (xs : List Int) ->
    traverse_ @Int @IO @List dict$Monad$IO dict$Foldable$List print (isort xs)
main$ll2 : Int -> IO Unit =
  fun (n : Int) ->
    bind @IO dict$Monad$IO @(List Int) @Unit (sequence @Int @IO dict$Monad$IO (replicate @(IO Int) n input)) main$ll1
main : IO Unit = bind @IO dict$Monad$IO @Int @Unit input main$ll2
