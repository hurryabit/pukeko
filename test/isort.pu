import std2/prelude

insert :: Int -> List Int -> List Int
insert y xs =
  match xs with
  | Nil -> Cons y Nil
  | Cons x xs' ->
    if y <= x then
      Cons y xs
    else
      Cons x (insert y xs')

isort :: List Int -> List Int
isort xs =
  match xs with
  | Nil -> Nil
  | Cons x xs -> insert x (isort xs)

main :: IO Unit
main =
  input
  >>= \n ->
  sequence (replicate n input)
  >>= \xs ->
  traverse_ print (isort xs)
