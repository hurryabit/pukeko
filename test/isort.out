192262039210

Steps  :   5806632471
Allocs :    124641340          (Checks:     99702071)
GC runs:            7
        1.85 real         1.51 user         0.31 sys
